File: code/php/rest/server/src/Models/Task.php
<?php
declare(strict_types=1);

namespace TaskAPI\Models;

use JsonSerializable;
use Ramsey\Uuid\Uuid;

class Task implements JsonSerializable
{
    public const STATUS_PENDING = 'pending';
    public const STATUS_IN_PROGRESS = 'in_progress';
    public const STATUS_COMPLETED = 'completed';
    public const STATUS_CANCELLED = 'cancelled';
    public const STATUS_ON_HOLD = 'on_hold';
    
    public const PRIORITY_LOW = 'low';
    public const PRIORITY_MEDIUM = 'medium';
    public const PRIORITY_HIGH = 'high';
    public const PRIORITY_CRITICAL = 'critical';
    
    private string $id;
    private string $title;
    private string $description;
    private string $status;
    private string $priority;
    private array $tags;
    private string $createdBy;
    private ?string $assignedTo;
    private string $createdAt;
    private string $updatedAt;
    private ?string $dueDate;
    private ?string $completedAt;
    
    public function __construct(
        string $title,
        string $description = '',
        string $status = self::STATUS_PENDING,
        string $priority = self::PRIORITY_MEDIUM,
        array $tags = [],
        string $createdBy = 'system',
        ?string $assignedTo = null,
        ?string $dueDate = null
    ) {
        $this->id = Uuid::uuid4()->toString();
        $this->title = $title;
        $this->description = $description;
        $this->status = $status;
        $this->priority = $priority;
        $this->tags = $tags;
        $this->createdBy = $createdBy;
        $this->assignedTo = $assignedTo;
        $this->createdAt = date('c');
        $this->updatedAt = date('c');
        $this->dueDate = $dueDate;
        $this->completedAt = null;
    }
    
    public static function fromArray(array $data): self
    {
        $task = new self(
            $data['title'],
            $data['description'] ?? '',
            $data['status'] ?? self::STATUS_PENDING,
            $data['priority'] ?? self::PRIORITY_MEDIUM,
            $data['tags'] ?? [],
            $data['created_by'] ?? 'system',
            $data['assigned_to'] ?? null,
            $data['due_date'] ?? null
        );
        
        if (isset($data['id'])) {
            $task->id = $data['id'];
        }
        
        return $task;
    }
    
    public function getId(): string
    {
        return $this->id;
    }
    
    public function getTitle(): string
    {
        return $this->title;
    }
    
    public function setTitle(string $title): void
    {
        $this->title = $title;
        $this->updatedAt = date('c');
    }
    
    public function getDescription(): string
    {
        return $this->description;
    }
    
    public function setDescription(string $description): void
    {
        $this->description = $description;
        $this->updatedAt = date('c');
    }
    
    public function getStatus(): string
    {
        return $this->status;
    }
    
    public function setStatus(string $status): void
    {
        $this->status = $status;
        if ($status === self::STATUS_COMPLETED && $this->completedAt === null) {
            $this->completedAt = date('c');
        }
        $this->updatedAt = date('c');
    }
    
    public function getPriority(): string
    {
        return $this->priority;
    }
    
    public function setPriority(string $priority): void
    {
        $this->priority = $priority;
        $this->updatedAt = date('c');
    }
    
    public function getTags(): array
    {
        return $this->tags;
    }
    
    public function setTags(array $tags): void
    {
        $this->tags = $tags;
        $this->updatedAt = date('c');
    }
    
    public function getAssignedTo(): ?string
    {
        return $this->assignedTo;
    }
    
    public function setAssignedTo(?string $assignedTo): void
    {
        $this->assignedTo = $assignedTo;
        $this->updatedAt = date('c');
    }
    
    public function getDueDate(): ?string
    {
        return $this->dueDate;
    }
    
    public function setDueDate(?string $dueDate): void
    {
        $this->dueDate = $dueDate;
        $this->updatedAt = date('c');
    }
    
    public function getCreatedAt(): string
    {
        return $this->createdAt;
    }
    
    public function getUpdatedAt(): string
    {
        return $this->updatedAt;
    }
    
    public function getCompletedAt(): ?string
    {
        return $this->completedAt;
    }
    
    public function isValid(): bool
    {
        if (empty($this->title) || strlen($this->title) > 200) {
            return false;
        }
        
        if (!in_array($this->status, $this->getValidStatuses())) {
            return false;
        }
        
        if (!in_array($this->priority, $this->getValidPriorities())) {
            return false;
        }
        
        return true;
    }
    
    public function getValidationErrors(): array
    {
        $errors = [];
        
        if (empty($this->title)) {
            $errors[] = 'Title is required';
        } elseif (strlen($this->title) > 200) {
            $errors[] = 'Title must be 200 characters or less';
        }
        
        if (!in_array($this->status, $this->getValidStatuses())) {
            $errors[] = "Invalid status: {$this->status}";
        }
        
        if (!in_array($this->priority, $this->getValidPriorities())) {
            $errors[] = "Invalid priority: {$this->priority}";
        }
        
        return $errors;
    }
    
    public function update(array $data): void
    {
        if (isset($data['title'])) {
            $this->setTitle($data['title']);
        }
        
        if (isset($data['description'])) {
            $this->setDescription($data['description']);
        }
        
        if (isset($data['status'])) {
            $this->setStatus($data['status']);
        }
        
        if (isset($data['priority'])) {
            $this->setPriority($data['priority']);
        }
        
        if (isset($data['tags'])) {
            $this->setTags($data['tags']);
        }
        
        if (array_key_exists('assigned_to', $data)) {
            $this->setAssignedTo($data['assigned_to']);
        }
        
        if (array_key_exists('due_date', $data)) {
            $this->setDueDate($data['due_date']);
        }
    }
    
    public function matchesFilters(?string $status = null, ?string $assignedTo = null, ?string $tags = null): bool
    {
        if ($status !== null && $this->status !== $status) {
            return false;
        }
        
        if ($assignedTo !== null && $this->assignedTo !== $assignedTo) {
            return false;
        }
        
        if ($tags !== null) {
            $requiredTags = array_map('trim', explode(',', $tags));
            foreach ($requiredTags as $tag) {
                if (!in_array($tag, $this->tags)) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    public function getPriorityValue(): int
    {
        return match($this->priority) {
            self::PRIORITY_LOW => 1,
            self::PRIORITY_MEDIUM => 2,
            self::PRIORITY_HIGH => 3,
            self::PRIORITY_CRITICAL => 4,
            default => 2
        };
    }
    
    public static function getValidStatuses(): array
    {
        return [
            self::STATUS_PENDING,
            self::STATUS_IN_PROGRESS,
            self::STATUS_COMPLETED,
            self::STATUS_CANCELLED,
            self::STATUS_ON_HOLD
        ];
    }
    
    public static function getValidPriorities(): array
    {
        return [
            self::PRIORITY_LOW,
            self::PRIORITY_MEDIUM,
            self::PRIORITY_HIGH,
            self::PRIORITY_CRITICAL
        ];
    }
    
    public function jsonSerialize(): array
    {
        $data = [
            'id' => $this->id,
            'title' => $this->title,
            'description' => $this->description,
            'status' => $this->status,
            'priority' => $this->priority,
            'tags' => $this->tags,
            'created_by' => $this->createdBy,
            'created_at' => $this->createdAt,
            'updated_at' => $this->updatedAt
        ];
        
        if ($this->assignedTo !== null) {
            $data['assigned_to'] = $this->assignedTo;
        }
        
        if ($this->dueDate !== null) {
            $data['due_date'] = $this->dueDate;
        }
        
        if ($this->completedAt !== null) {
            $data['completed_at'] = $this->completedAt;
        }
        
        return $data;
    }
}

File: code/ruby/rest/server/models/task.rb
require 'json'
require 'time'
require 'securerandom'

module Models
  class Task
    VALID_STATUSES = %w[pending in_progress completed cancelled on_hold].freeze
    VALID_PRIORITIES = %w[low medium high critical].freeze
    
    attr_accessor :id, :title, :description, :status, :priority, :tags,
                  :created_by, :assigned_to, :created_at, :updated_at,
                  :due_date, :completed_at
    
    def initialize(attrs = {})
      @id = attrs[:id] || SecureRandom.uuid
      @title = attrs[:title] || ''
      @description = attrs[:description] || ''
      @status = attrs[:status] || 'pending'
      @priority = attrs[:priority] || 'medium'
      @tags = attrs[:tags] || []
      @created_by = attrs[:created_by] || 'system'
      @assigned_to = attrs[:assigned_to]
      @created_at = attrs[:created_at] || Time.now
      @updated_at = attrs[:updated_at] || Time.now
      @due_date = attrs[:due_date]
      @completed_at = attrs[:completed_at]
    end
    
    def self.from_json(json_str)
      data = JSON.parse(json_str, symbolize_names: true)
      from_hash(data)
    end
    
    def self.from_hash(data)
      task = new(data)
      
      # Parse time fields
      task.created_at = Time.parse(data[:created_at]) if data[:created_at].is_a?(String)
      task.updated_at = Time.parse(data[:updated_at]) if data[:updated_at].is_a?(String)
      task.due_date = Time.parse(data[:due_date]) if data[:due_date].is_a?(String)
      task.completed_at = Time.parse(data[:completed_at]) if data[:completed_at].is_a?(String)
      
      task
    end
    
    def to_h
      {
        id: @id,
        title: @title,
        description: @description,
        status: @status,
        priority: @priority,
        tags: @tags,
        created_by: @created_by,
        assigned_to: @assigned_to,
        created_at: @created_at.iso8601,
        updated_at: @updated_at.iso8601,
        due_date: @due_date&.iso8601,
        completed_at: @completed_at&.iso8601
      }.compact
    end
    
    def to_json(*args)
      to_h.to_json(*args)
    end
    
    def valid?
      return false if @title.nil? || @title.empty?
      return false if @title.length > 200
      return false unless VALID_STATUSES.include?(@status)
      return false unless VALID_PRIORITIES.include?(@priority)
      true
    end
    
    def validation_errors
      errors = []
      errors << "Title is required" if @title.nil? || @title.empty?
      errors << "Title must be 200 characters or less" if @title && @title.length > 200
      errors << "Invalid status: #{@status}" unless VALID_STATUSES.include?(@status)
      errors << "Invalid priority: #{@priority}" unless VALID_PRIORITIES.include?(@priority)
      errors
    end
    
    def update!(attrs)
      @title = attrs[:title] if attrs.key?(:title)
      @description = attrs[:description] if attrs.key?(:description)
      @status = attrs[:status] if attrs.key?(:status)
      @priority = attrs[:priority] if attrs.key?(:priority)
      @tags = attrs[:tags] if attrs.key?(:tags)
      @assigned_to = attrs[:assigned_to] if attrs.key?(:assigned_to)
      @due_date = attrs[:due_date] if attrs.key?(:due_date)
      
      # Update completed_at if status changes to completed
      if @status == 'completed' && @completed_at.nil?
        @completed_at = Time.now
      end
      
      @updated_at = Time.now
      self
    end
    
    def matches_filters?(status: nil, assigned_to: nil, tags: nil)
      return false if status && @status != status
      return false if assigned_to && @assigned_to != assigned_to
      
      if tags && !tags.empty?
        tag_list = tags.is_a?(String) ? tags.split(',').map(&:strip) : tags
        return false unless (tag_list - @tags).empty?
      end
      
      true
    end
    
    def priority_value
      case @priority
      when 'low' then 1
      when 'medium' then 2
      when 'high' then 3
      when 'critical' then 4
      else 2
      end
    end
  end
end